### PR TITLE
manifest, push: implement `--add-compression` to push with compressed variants.

### PR DESCRIPTION
**Pull request by flouthoc**
This is a feature PR for containers/image feature merged in below PR:
* https://github.com/containers/image/pull/2023
* https://github.com/containers/image/pull/1987
* https://github.com/containers/image/pull/2029

Adds a new flag  **--add-compression <compression>**  to manifest push, which when added will push existing manifest on host while replicating the instances with required compression.

```release-note
manifest, push: implement --add-compression to push with compressed variants.
```

Similar implementation to: https://github.com/containers/buildah/pull/4912

### CODE REVIEW

**Comment by flouthoc** [comment]
Needs: https://github.com/containers/common/pull/1590#pullrequestreview-1556349149
/hold

**Comment by flouthoc** [comment]
@containers/podman-maintainers PTAL PR is similar to https://github.com/containers/buildah/pull/4912

**Comment by flouthoc** [comment]
@mtrmac PTAL

File: docs/source/markdown/podman-manifest-push.1.md.in
Comment (mtrmac) [inline docs/source/markdown/podman-manifest-push.1.md.in:22]: That option is named `--compression-format` in Podman.

File: pkg/api/handlers/libpod/manifests.go
Comment (mtrmac) [inline pkg/api/handlers/libpod/manifests.go:343]:
```suggestion
		AddCompression    []string `schema:"addCompression"`
```
?

File: pkg/domain/entities/images.go
Comment (mtrmac) [inline pkg/domain/entities/images.go:247]: Do we need to worry about `podman push` (i.e. not a `manifest push`) with `AddCompression` set?

(That’s the price of sharing the same `ImagePushOptions` for two operations…)

I can see the argument that the only reasonable callers are from inside the Podman codebase, so we don’t care…

Still, probably worth at least adding a comment that this only applies to manifest push, _for now_. (I think we will want to add that in the future.)

File: pkg/domain/infra/abi/manifest.go
Comment (mtrmac) [inline pkg/domain/infra/abi/manifest.go:167]: If I understand this correctly, the semantics `if false` is not “OCI” but “raw without conversion”. Why name the format value OCI, then?

File: cmd/podman/manifest/inspect.go
Comment (mtrmac) [inline cmd/podman/manifest/inspect.go:41]: As is, I’m rather unhappy with adding an option
- where the actual semantics (“raw”) differs from the claimed one (“OCI”)
- with such a generic name, with a fair risk of someone external using it (`--internal-unstable-output-raw-format` would work better)
- which _silently_ does nothing for remote clients

Would be possible to run `skopeo inspect --raw` instead of adding this partially-implemented option (and instead of taking the effort to implement it fully)?

At least _some_ parts of `test/system` seem to depend on the Skopeo binary already.

**Review by mtrmac — COMMENTED** [review]
Podman maintainers, please take a look as well — I think I have a general understanding but there are certainly aspects of the engine / remote API design that I’m missing.

File: docs/source/markdown/podman-manifest-push.1.md.in
Comment (rhatdan) [reply docs/source/markdown/podman-manifest-push.1.md.in:22]: Can we just extend --compression to be a comma separated list.

--compression gzip,zstd

I think that makes for a better User UI.

Comment (mtrmac) [reply docs/source/markdown/podman-manifest-push.1.md.in:22]: Right now the two have quite different semantics, as the text above says.

Comment (rhatdan) [reply docs/source/markdown/podman-manifest-push.1.md.in:22]: Not sure why you say that.  
If I say --compression gzip, then it is going to push the image if it has gzip compression or change the compression to gzip if it was something else.
If I push --compression gzip,zstd then I would figure it is going to push gzip and zstd, even if the original was something different.

File: cmd/podman/manifest/inspect.go
Comment (flouthoc) [reply cmd/podman/manifest/inspect.go]: Done using skopeo now.

File: pkg/domain/infra/abi/manifest.go
Comment (flouthoc) [reply pkg/domain/infra/abi/manifest.go]: Not using this, instead using `skopeo` in test.

File: docs/source/markdown/podman-manifest-push.1.md.in
Comment (flouthoc) [reply docs/source/markdown/podman-manifest-push.1.md.in]: @rhatdan We have already used `--add-compression` since it was overlapping with `--compression` on buildah side, it will be good here for consistency.  

Extending `--compression` changes functionality for example `--compression gzip` will replace instances , but `--compression gzip,zstd` will only replace `gzip` and add `zstd` while changing the order of input will change functionality entirely  consider `--compression zstd,gzip` it will replace `zstd` and add `gzip`. 

We had a prior discussion here: https://github.com/containers/common/pull/1585

File: pkg/api/handlers/libpod/manifests.go
Comment (flouthoc) [reply pkg/api/handlers/libpod/manifests.go]: Done added and tested for `podman-remote` manually.

File: docs/source/markdown/podman-manifest-push.1.md.in
Comment (mtrmac) [reply docs/source/markdown/podman-manifest-push.1.md.in]: The _existing_ `--compress` prefers to reuse blobs, even if they don’t use the expected algorithm. That’s a long-standing behavior and I wouldn’t be surprised if users by now relied on it.

(It’s not necessarily the option semantics we would design _now_ as the default, if we were building the CLI from scratch with the feature set of the underlying codebase as of today, but I think the years of past practice change the calculation. @flouthoc is already working on making it possible to force using the requested algorithm independently from the cloning functionality, but that’s not _this_ PR.)

File: pkg/bindings/manifests/manifests.go
Comment (mtrmac) [inline pkg/bindings/manifests/manifests.go:233]: `options.ToParams` doesn’t handle this? As the comment above says, `SkipTLSVerify` is special, and all of the other options in `options` are not added individually in this function.

(Note that I didn’t try this in practice, and I didn’t even read `ToParams` carefully, so I don’t know what I don’t know.)

File: pkg/domain/entities/images.go
Comment (mtrmac) [inline pkg/domain/entities/images.go:247]: maybe 
```suggestion
	// If necessary, add clones of existing instances with requested compression algorithms to manifest list
```
?

OTOH as is, this matches the c/common comment

**Review by mtrmac — COMMENTED** [review]
Thanks, this looks close to done to me.

At this point I’ll need to defer to actual Podman experts — please review independently.

File: pkg/bindings/manifests/manifests.go
Comment (flouthoc) [reply pkg/bindings/manifests/manifests.go]: Corrent `ToParams` handles it internally as well, removed this.

File: pkg/domain/entities/images.go
Comment (flouthoc) [reply pkg/domain/entities/images.go]: done.

**Comment by flouthoc** [comment]
Rebased.

File: cmd/podman/manifest/push.go
Comment (Luap99) [inline cmd/podman/manifest/push.go:61]: This autocompletion function makes no sense, it will suggest file paths.

Assuming it uses the some formats as --compression-format you can just reuse `common.AutocompleteCompressionFormat`

File: pkg/api/handlers/libpod/manifests.go
Comment (Luap99) [inline pkg/api/handlers/libpod/manifests.go:343]: this would need an update to the swagger doc to list the new option?

File: test/system/012-manifest.bats
Comment (Luap99) [inline test/system/012-manifest.bats:95]: to my knowledge skopeo is a hard requirement so this skip makes no sense
cc @edsantiago

File: test/system/012-manifest.bats
Comment (Luap99) [inline test/system/012-manifest.bats:96]: why? Not having remote tests is a big issue. We need tests to cover remote client and API as well.
If we cannot set this up in system tests we should use e2e tests or the bindings and API tests.

File: test/system/012-manifest.bats
Comment (flouthoc) [reply test/system/012-manifest.bats:95]: The tests which are using `skopeo` have this check directly or indirectly so I have added it here.

File: pkg/api/handlers/libpod/manifests.go
Comment (flouthoc) [reply pkg/api/handlers/libpod/manifests.go:343]: done.

File: cmd/podman/manifest/push.go
Comment (flouthoc) [reply cmd/podman/manifest/push.go]: done.

File: test/system/012-manifest.bats
Comment (flouthoc) [reply test/system/012-manifest.bats:96]: Added a `e2e` test.

File: test/e2e/manifest_test.go
Comment (mtrmac) [inline test/e2e/manifest_test.go:201]: Absolutely non-blocking: This could plausibly call c/image instead of using a subprocess.

OTOH this is easier to follow for people unfamiliar with that codebase.

File: test/e2e/manifest_test.go
Comment (mtrmac) [inline test/e2e/manifest_test.go:182]: Do we have multi-arch image that could be used for this test without requiring a build, by any chance?

(I have no idea … and if this works, maybe we can spend CPU cycles instead of our attention.)

**Review by mtrmac — COMMENTED** [review]
Absolutely non-blocking tentative suggestions — don’t let me slow down this work unnecessarily.

File: test/e2e/manifest_test.go
Comment (edsantiago) [reply test/e2e/manifest_test.go:182]: `quay.io/libpod/20221018`?  Or even `:00000004` if you want a tiny image.

File: test/e2e/manifest_test.go
Comment (flouthoc) [reply test/e2e/manifest_test.go:201]: This is similar to systems test for easier review and familiarity.

File: test/e2e/manifest_test.go
Comment (flouthoc) [reply test/e2e/manifest_test.go:182]: Sure I'll consider it if I am going to make any further change and repush this. ( I hope this not a hard blocker since OTOH building image makes tests independent of any external image ( pulling ) which has been encouraged in past if i can recall )

File: test/e2e/manifest_test.go
Comment (Luap99) [reply test/e2e/manifest_test.go:201]: I think it would make more sense to directly execute skopeo on the host instead of pulling the image as pull are flaky and slow. We already depend on it in system test so I see no problem using it in e2e as well.

**Review by Luap99 — APPROVED** [review]
LGTM, tests can be improved later if needed.

File: docs/source/markdown/podman-manifest-push.1.md.in
Comment (vrothberg) [inline docs/source/markdown/podman-manifest-push.1.md.in:23]: For some reason I cannot propose a change via the GitHub UI but the sentence above has typos:

makes sure tha**t** each instance has **its** variant added to manifest list without modifying the
original instance.

I still find the note hard to understand and suggest to be more explicit and less abstract:

"Note that `--compression-format` controls the compression format of each instance in the manifest list.  `--add-compression` will add another variant for each instance in the list with the specified compressions.  `--compression-format gzip --add-compression zstd` will push a manifest list with each instance being compressed with gzip plus an additional variant of each instance being compressed with zstd."

I am sure there's a better description than mine but I find it important to be more explicit.

**Review by vrothberg — COMMENTED** [review]
One last nit on the docs

File: docs/source/markdown/podman-manifest-push.1.md.in
Comment (vrothberg) [reply docs/source/markdown/podman-manifest-push.1.md.in:23]: Maybe worth mentioning that it can be specified multiple times.

Comment (flouthoc) [reply docs/source/markdown/podman-manifest-push.1.md.in]: > Maybe worth mentioning that it can be specified multiple times.

Amended text and added above line as well.

**Review by vrothberg — APPROVED** [review]
/lgtm
/hold

**Comment by flouthoc** [comment]
cc @TomSweeneyRedHat

File: test/e2e/manifest_test.go
Comment (mtrmac) [reply test/e2e/manifest_test.go:182]: IIRC the Podman e2e tests already have some infrastructure to pre-pull / cache images, so that not _every_ pull needs to reach out to the internet. But I don’t know the details by heart.

**Pull request merged**
